### DISCUSSION TITLE
Help getting started with snippets: pointer to "best practices-following instructions"?

### DISCUSSION THREAD

**Discussion by jerabaul29** [helix-editor/helix#9349] [Q&A]
I understand that there is some form of support for snippets in helix, but I am a bit confused; see for example:

- though this issue is open: https://github.com/helix-editor/helix/issues/395
- snippets are mentioned here, if I understand well it is recommended to use a LSP to provide for this? : https://helix-editor.com/news/release-23-03-highlights/ . There is also this issue that was merged: https://github.com/helix-editor/helix/pull/5864 . I can find for example:
  - https://github.com/estin/simple-completion-language-server
  - https://github.com/quantonganh/snippets-ls
  - https://github.com/rajasegar/helix-snippets-ls

I am also confused by https://github.com/helix-editor/helix/discussions/6449 ; does this mean that helix has now some form of built in snippet LSP, or did the user mean that they have installed a separate snippet LSP, then added this to their config?

My questions:

- is there an up to date "getting started guide following current helix best practices" about how to use snippets together with helix somewhere? I looked at the documentation but could not find some.
- is everything needed to use snippets built in, or is it necessary to install a separate / external LSP server? If so, can you point to what LSP server should be installed / any "default recommendation"?

**Comment by the-mikedavis** [comment]
We currently support LSP snippets, see https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax. So if a language server sends us a completion in snippet format then we can apply it. #395 covers custom snippets, for example in vscode see https://code.visualstudio.com/docs/editor/userdefinedsnippets. Those aren't implemented in Helix yet so there are some language servers you can use as workarounds like you linked. I don't use custom snippets so I can't recommend any of them over the others.

**Reply by jerabaul29** [reply]
Thanks for the explanation. Just a couple of questions: i) do you think helix will implement snippets "natively" in the future (with language dependent list of snippets etc), or do you think this will always require an external "snippets lsp"? ii) is there a way, given a snippet provided by an external lsp, to jump between snippet pre-defined fields?

**Reply by the-mikedavis** [reply]
IMO it would be ok to support user/custom snippets in core (i.e. not a plugin) since we already need all the snippet parsing and application code for LSP. I think it would be a medium/large effort to add it though since we currently only use LSP for completions (also see #1015).

Jumping between tabstops isn't currently implemented: it was left off of #5864 since it would take large changes to add it and that PR was already quite large. It should be implemented in the future though and usable for snippets from language servers as well as custom snippets.
